File: apps/web/lib/types/analytics.ts
export type WeakConcept = {
  concept_id: number;
  slug: string;
  name: string;
  topic_id: number;
  topic_slug: string;
  mastery_score: number;
  attempts_count: number;
};

export type AttemptHistoryItem = {
  id: number;
  question_id: number;
  question_title: string;
  topic_id: number;
  topic_slug: string;
  is_correct: boolean | null;
  score: number | null;
  time_spent_seconds: number;
  created_at: string;
};

export type SearchMiss = {
  id: number;
  query: string;
  topic_slug: string | null;
  types: string[];
  created_at: string;
};

export type LearningAnalytics = {
  user_id: string;
  weak_concepts: WeakConcept[];
  attempt_history: AttemptHistoryItem[];
  review_due_count: number;
  search_misses: SearchMiss[];
};

File: apps/api/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel, Field


class WeakConceptRead(BaseModel):
    concept_id: int
    slug: str
    name: str
    topic_id: int
    topic_slug: str
    mastery_score: float
    attempts_count: int


class AttemptHistoryItemRead(BaseModel):
    id: int
    question_id: int
    question_title: str
    topic_id: int
    topic_slug: str
    is_correct: bool | None
    score: float | None
    time_spent_seconds: int
    created_at: datetime


class SearchMissRead(BaseModel):
    id: int
    query: str
    topic_slug: str | None = None
    types: list[str] = Field(default_factory=list)
    created_at: datetime


class LearningAnalyticsRead(BaseModel):
    user_id: str
    weak_concepts: list[WeakConceptRead]
    attempt_history: list[AttemptHistoryItemRead]
    review_due_count: int
    search_misses: list[SearchMissRead]
